src/server/system_services/bucket_server.js

```diff
@@ -489,26 +489,46 @@ async function put_bucket_policy(req) {
 function _validate_s3_policy(policy, bucket_name) {
     const all_op_names = _.compact(_.flatMap(OP_NAME_TO_ACTION, action => [action.regular, action.versioned]));
     for (const statement of policy.Statement) {
-        if (statement.Principal.AWS) {
-            for (const principal of _.flatten([statement.Principal.AWS])) {
+        if (!statement.Principal && !statement.NotPrincipal) {
+            throw new RpcError('MALFORMED_POLICY', 'must have Principal or NotPrincipal in policy', { detail: "" });
+        }
+        if (statement.Principal && statement.NotPrincipal) {
+            throw new RpcError('MALFORMED_POLICY', 'The policy elements Principal and NotPrincipal cannot be used in the same statement', { detail: "" });
```

Comment:
Why do you send `{ detail: "" }` to the `RpcError`?
I think you don't have to:
https://github.com/noobaa/noobaa-core/blob/c497a4f17104081b58bcafbf5b64979cbb0f9ebc/src/rpc/rpc_error.js#L21-L23

Reply:
I have to for MALFORMED_POLICY errors. because this value is used in 
https://github.com/noobaa/noobaa-core/blob/c497a4f17104081b58bcafbf5b64979cbb0f9ebc/src/endpoint/s3/s3_rest.js#L369-L372

Reply:
You created `RpcError` with property `rpc_code` (not `code`) of `MALFORMED_POLICY` (not `MalformedPolicy`). If you can explain in the flow what will change from where we start to the code snippet you shared.

Reply:
not relevant anymore after removal, but we need it in case of MALFORMED_POLICY:
(https://github.com/noobaa/noobaa-core/blob/c497a4f17104081b58bcafbf5b64979cbb0f9ebc/src/endpoint/s3/s3_rest.js#L369-L372)